src/server.rs

```diff
@@ -1,10 +1,154 @@
-use std::net::SocketAddr;
+use types::*;
+use Application;
 
-use super::Application;
+use std::io;
+use std::io::*;
+use std::net::*;
+use std::ops::DerefMut;
+use std::sync::{Arc, Mutex};
+use std::thread;
 
-use tcpserver::*;
+use bytes::{BufMut, BytesMut};
+use integer_encoding::VarInt;
 
-pub fn new<A: Application + Send + Sync + 'static>(listen_addr: SocketAddr, app: &'static A) {
-    let mut server = TCPServer::new(app, listen_addr);
-    server.serve().unwrap();
-}
\ No newline at end of file
+use protobuf;
+use protobuf::Message;
+
+const BUFFER_SIZE: usize = 4096;
+
+// Creates the TCP server and listens for connections from Tendermint
+pub fn serve<A>(app: A, addr: SocketAddr) -> io::Result<()>
+where
+    A: Application + 'static + Send + Sync,
+{
+    let listener = TcpListener::bind(addr).unwrap();
+
+    // Wrap the app atomically and clone for each connection.
+    //
+    // The Mutex is probably not *really* needed as Tendermint synchronizes calls...just being
+    // safe in the event that changes.
+    let app = Arc::new(Mutex::new(app));
```

Comment:
Just to correct your comment here, we actually do need a mutex, for internal mutability. We could use a `RefCell` also, but a `Mutex` will be fine for our uses because we don't need to worry about deadlocks.